header/p2p/session.go

```diff
@@ -130,17 +134,17 @@ func (s *session) doRequest(
 	req *p2p_pb.ExtendedHeaderRequest,
 	headers chan []*header.ExtendedHeader,
 ) {
+	ctx, cancel := context.WithTimeout(ctx, s.requestTimeout)
+	defer cancel()
+
 	r, size, duration, err := sendMessage(ctx, s.host, stat.peerID, s.protocolID, req)
 	if err != nil {
-		if err == context.Canceled || err == context.DeadlineExceeded {
-			return
-		}
 		log.Errorw("requesting headers from peer failed.", "failed peer", stat.peerID, "err", err)
 		select {
-		case <-ctx.Done():
 		case <-s.ctx.Done():
 			// retry request
```

Comment:
I don't understand how this leads to a retry of the request?

Reply:
as ctx.Done is removed, we will fall in `case s.reqCh <- req:` (s.reqCh is a buffered channel, so we will not block there)

Reply:
the comment is confusing then, let's move it